src/analyze/analyze-shutdown-blame.c

```diff
@@ -0,0 +1,258 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to add systemd process match: %m");
+
+        r = sd_journal_add_conjunction(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add journal conjunction: %m");
+
+        /* Match unit stopping messages */
+        r = sd_journal_add_disjunction(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add journal disjunction: %m");
+
+        r = sd_journal_add_match(j, "MESSAGE_ID=" SD_MESSAGE_UNIT_STOPPING_STR, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add stopping message match: %m");
+
+        r = sd_journal_add_match(j, "MESSAGE_ID=" SD_MESSAGE_UNIT_STOPPED_STR, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add stopped message match: %m");
+
+        /* Seek to the end and go backwards to find the most recent shutdown */
+        r = sd_journal_seek_tail(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to journal tail: %m");
+
+        r = sd_journal_previous(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to iterate journal: %m");
+
+        SD_JOURNAL_FOREACH_BACKWARDS(j) {
+                const char *message_id = NULL, *unit = NULL;
+                usec_t timestamp;
+                _cleanup_(shutdown_time_freep) ShutdownTime *st = NULL;
+                ShutdownTime *existing;
+                bool is_stopping;
+
+                r = sd_journal_get_realtime_usec(j, &timestamp);
+                if (r < 0)
+                        continue;
+
+                r = sd_journal_get_data(j, "MESSAGE_ID", &data, &length);
+                if (r < 0)
+                        continue;
+
+                message_id = (const char*) data + STRLEN("MESSAGE_ID=");
+                if (length <= STRLEN("MESSAGE_ID="))
+                        continue;
+
+                r = sd_journal_get_data(j, "UNIT", &data, &length);
+                if (r < 0)
+                        continue;
+
+                unit = (const char*) data + STRLEN("UNIT=");
+                if (length <= STRLEN("UNIT="))
+                        continue;
+
+                /* Check if this is a stopping or stopped message */
+                if (startswith(message_id, SD_MESSAGE_UNIT_STOPPING_STR))
+                        is_stopping = true;
+                else if (startswith(message_id, SD_MESSAGE_UNIT_STOPPED_STR))
+                        is_stopping = false;
+                else
+                        continue;
+
+                existing = hashmap_get(shutdown_times, unit);
+
+                if (is_stopping) {
+                        if (existing) {
+                                existing->start_time = timestamp;
+                                if (existing->stop_time > 0)
+                                        existing->duration = existing->stop_time - existing->start_time;
```

Comment:
maybe check first if stop time is really larger thatn start time, before doing the substraction, just to be safe.